crates/dbsp/src/trace/layers/column_layer/builders.rs

```diff
@@ -167,6 +194,85 @@ where
         unsafe { self.assume_invariants() }
         self.keys.len()
     }
+
+    fn push_merge_retain_keys<'a, F>(
+        &'a mut self,
+        cursor1: <Self::Trie as Trie>::Cursor<'a>,
+        cursor2: <Self::Trie as Trie>::Cursor<'a>,
+        filter: &F,
+    ) -> usize
+    where
+        F: Fn(&<<Self::Trie as Trie>::Cursor<'a> as Cursor<'a>>::Key) -> bool,
+    {
+        unsafe { self.assume_invariants() }
+
+        let (trie1, trie2) = (cursor1.storage(), cursor2.storage());
+        unsafe {
+            trie1.assume_invariants();
+            trie2.assume_invariants();
+        }
+
+        let (_, upper1) = cursor1.bounds();
+        let mut lower1 = cursor1.position();
+        let (_, upper2) = cursor2.bounds();
+        let mut lower2 = cursor2.position();
+
+        let reserved = (upper1 - lower1) + (upper2 - lower2);
+        self.reserve(reserved);
+
+        // while both mergees are still active
+        while lower1 < upper1 && lower2 < upper2 {
+            match trie1.keys[lower1].cmp(&trie2.keys[lower2]) {
+                Ordering::Less => {
+                    // determine how far we can advance lower1 until we reach/pass lower2
+                    let step = 1 + advance(&trie1.keys[(1 + lower1)..upper1], |x| {
+                        x < &trie2.keys[lower2]
+                    });
+
+                    let step = min(step, 1000);
```

Comment:
what's this for?

Reply:
heuristic I inherited from DD.

Reply:
maybe you can make 1000 a constant somewhere